### PR TITLE
Don't throw in cholfact if matrix is not Hermitian

### PR DESCRIPTION
**Pull request by fredrikekre**
As discussed [here](https://github.com/JuliaLang/LinearAlgebra.jl/issues/182) and comments below.

We could perhaps dedicate an info code for non-hermitianness, like `-1` or something instead? Then we could throw a more targeted error message (similar to what we had before this PR) or throw a `HermitianException`, instead of
```julia
julia> A = rand(5,5); x = rand(5); cholfact(A)\x
ERROR: Base.LinAlg.PosDefException(1)
Stacktrace:
 [1] A_ldiv_B! at ./linalg/cholesky.jl:412 [inlined]
 [2] \(::Base.LinAlg.Cholesky{Float64,Array{Float64,2}}, ::Array{Float64,1}) at ./linalg/factorization.jl:71
```

Thoughts?

### CODE REVIEW

**Comment by KristofferC** [comment]
Make an Enum for info codes?

File: test/linalg/cholesky.jl [hunk 1]
```diff
@@ -261,14 +261,14 @@ end
...
-@testset "throw if non-Hermitian" begin
+@testset "handling of non-Hermitian" begin
     R = randn(5, 5)
     C = complex.(R, R)
     for A in (R, C)
-        @test_throws ArgumentError cholfact(A)
-        @test_throws ArgumentError cholfact!(copy(A))
-        @test_throws ArgumentError chol(A)
-        @test_throws ArgumentError Base.LinAlg.chol!(copy(A))
+        @test cholfact(A).info == 1
```
Comment (andreasnoack) [inline test/linalg/cholesky.jl]: Maybe test with `!issuccess` instead.

**Review by andreasnoack — APPROVED** [review]
In LAPACK, a positive info code corresponds to the index of the non-positive diagonal entry. However, we don't use that information right now so it is not critical. We could use a negative error code because they are only used in LAPACK when the input is invalid so that should never happen.

**Comment by andreasnoack** [comment]
> Make an Enum for info codes?

Right now the code encodes where the factorization failed. Would you just have two values? One for non-Hermitian and one for non-positive pivot?

File: test/linalg/cholesky.jl (hunk 1 above)
Comment (fredrikekre) [inline test/linalg/cholesky.jl]: Yea, that's better.

**Comment by fredrikekre** [comment]
Right, since a negative info code will error before we exit the LAPACK module, we could use a negative code for non-hermitian. And add a `show` method for `PosDefException`, and, in case of `info = -1` print something like: `ERROR: Base.LinAlg.PosDefException(3): matrix is not Hermitian.` Thoughts on that? :)

File: base/linalg/exceptions.jl [hunk 2]
```diff
@@ -32,6 +32,15 @@ end
 struct PosDefException <: Exception
     info::BlasInt
 end
+function Base.showerror(io::IO, ex::PosDefException)
+    print(io, "PosDefException: matrix is not ")
```
Comment (fredrikekre) [inline base/linalg/exceptions.jl]: `ERROR: PosDefException: matrix is not Hermitian.` is maybe not the nicest though. Should we instead have a `CholeskyException` that print one of 
```
ERROR: CholeskyException: Cholesky factorization failed; matrix is not Hermitian.
ERROR: CholeskyException: Cholesky factorization failed; matrix is not positive definite.
```

Comment (andreasnoack) [inline base/linalg/exceptions.jl]: I think `PosDef` is fine. We use the definition that requires the matrix to be `Hermitian` consistently. Some users have been taught the other definition that allows for non-Hermitian matrices but I think the error message is sufficiently clear about the reason for the failure.

Comment (fredrikekre) [inline base/linalg/exceptions.jl]: So, as the PR currently? Or should I drop the second commit?

**Comment by andreasnoack** [comment]
No. I think it is fine as it is.

**Pull request merged**